rlog/2024-12-23-merkle-tree.mdx

```diff
@@ -0,0 +1,213 @@
+|$\vdots$|
+|$d_{n-1} := \mathsf{Hash(d_{n-2} \|\| d_{n-2})}.$|
+
+Each of the $d_i$'s represents the root hash of a Merkle tree with $2^i$ leaves containing $\mathsf{null}$.
+These values can be used to shorten the time needed to construct an SMT and the length of proofs.
+
+### Proof of nonmembership
+
+In the first Vac 101 [[5](https://vac.dev/rlog/vac101-membership-with-bloom-filters-and-cuckoo-filters)], we examined Bloom and Cuckoo filters that could be used for proof of membership and nonmembership.
+However, the proof of membership may result in false positives due to collisions.
+This would affect nonmembership proofs as well.
+Sparse Merkle trees can be adapted to provide greater assurance that a given piece of data is not a member of the digest.
+
+Why is sorting essential?
+The sorting mechanism of data can be arbitrarily chosen.
+However, it is essential that there are no gaps in the ordering.
+The maximum number of elements that could ever exist in the digest must be known.
+A simple method for this is to use a hash function to provide fingerprints to the data.
+Each hash using either SHA-256 or Keccak has 256-bits.
+Our entire digest could consist of a maximum of $2^{256}$ entries.
+This assumes that our digest does not contain collisions.
+
+The fingerprint of a piece of data $\ell$ indicates which leaf of the SMT it is contained in.
+This means that a nonmembership of $\ell$ in the SMT becomes a matter of proving that $\mathsf{null}$ is contained in $\ell$'s location. 
+
+It is crucial for the SMT to be sorted. 
+Otherwise, a malicious party can append the entry $\ell$ to a random location.
+This allows for the malicious party to provide contradictory proofs that prove both membership and nonmembership.
+We note that the requirement that an SMT is sorted can be too strong of an assumption in centralized cases.
+However, sorted is a necessary property of SMTs for decentralized systems.
+
+### Verkle Trees
+A proof of membership grows in length as the Merkle tree grows.
+The most obvious approach to remedy this scalability issue is to use Merkle trees in which each node has more than two children.
+However, this does not fix the issue.
+A proof of membership in a $k$-nary Merkle tree [[6](https://math.mit.edu/research/highschool/primes/materials/2018/Kuszmaul.pdf)] (each node has $k$ children) has a proof size $\log_k(n)(k-1)$.
+The multiple $k-1$ is the number of silbings that a node has on each layer.
+Hence, the complexity for scalability is larger than a logarithmic function.
+
+An alternate approach is to use a different data structure: Verkle trees [[6](https://math.mit.edu/research/highschool/primes/materials/2018/Kuszmaul.pdf)].
+A Verkle tree replaces hash functions with polynomial commitments [[7](https://ethresear.ch/t/using-polynomial-commitments-to-replace-state-roots/7095), [8](https://dankradfeist.de/ethereum/2020/06/16/kate-polynomial-commitments.html)].
+We will explore Verkle trees is a future Vac 101 edition.
```

Comment:
Typo: "is" instead of "in".

Reply:
This will be a future blog. To describe Verkle trees, we have to assume knowledge of polynomial commitments. Which will be a Vac 101 between Merkle trees and Verkle trees.